### PR TITLE
Login register

### PR DESCRIPTION
**Pull request by TheBestTvarynka**
I had implemented the Login/Register flow on both Frontend and Backend sides.
Short description:
For authentication, I use sessions (client stores session id in cookies). By default Spring stores sessions in the HTTP session of the underlying webserver (Tomcat in our case). But this behavior is not suitable for me. Reasons:
1) If the backend server crashes then all my sessions will be deleted.
2) I would like to control flow by myself

SO, how it works? Check comments below for more details.

### CODE REVIEW

File: backend/src/main/java/com/eshka/config/WebConfig.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+@Configuration
+@EnableWebMvc
+public class WebConfig implements WebMvcConfigurer {
+
+    @Override
+    public void addCorsMappings(CorsRegistry registry) {
+        registry.addMapping("/**")
+		.allowedMethods("PUT", "GET", "DELETE", "OPTIONS", "PATCH", "POST")
+		.allowCredentials(true);
+    }
```
Comment (TheBestTvarynka) [inline backend/src/main/java/com/eshka/config/WebConfig.java:17]: Here I allow Cross-Origin Requests.

**Review by KySo4oK — DISMISSED** [review]

File: backend/src/main/java/com/eshka/config/security/AuthCookieFilter.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import com.eshka.entity.UserSession;
+import com.eshka.repository.UserSessionRepository;
+import lombok.extern.log4j.Log4j2;
+import org.springframework.http.HttpStatus;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.web.filter.GenericFilterBean;
+
+@Log4j2
+public class AuthCookieFilter extends GenericFilterBean {
+    public final static String COOKIE_NAME = "Authentication";
```
Comment (TheBestTvarynka) [inline backend/src/main/java/com/eshka/config/security/AuthCookieFilter.java:23]: This class filters incoming requests based on cookies.

File: backend/src/main/java/com/eshka/config/security/SecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                .authorizeRequests(cust -> {
+                    cust.antMatchers("/auth/login", "/auth/register")
+                            .permitAll()
+                            .anyRequest()
+                            .authenticated();
+                })
+                .exceptionHandling(cust -> cust
+                        .authenticationEntryPoint(new HttpStatusEntryPoint(HttpStatus.UNAUTHORIZED)))
+                .addFilterAfter(this.authCookieFilter, SecurityContextPersistenceFilter.class);
+    }
```
Comment (TheBestTvarynka) [inline backend/src/main/java/com/eshka/config/security/SecurityConfig.java:74]: Custom setting for security.

File: backend/src/main/java/com/eshka/config/security/SecurityConfig.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    @Override
+    public void onLogoutSuccess(HttpServletRequest request, HttpServletResponse response,
+                                Authentication authentication) throws IOException {
+        String sessionId = AuthCookieFilter.extractAuthenticationCookie(request);
+        sessionRepository.deleteById(sessionId);
+
+        response.setStatus(HttpServletResponse.SC_OK);
+        response.getWriter().flush();
+    }
+}
```
Comment (TheBestTvarynka) [inline backend/src/main/java/com/eshka/config/security/SecurityConfig.java:94]: When the user logs out I just delete the session from the database.

File: backend/src/main/java/com/eshka/controller/UserController.java [hunk 5]
```diff
@@ -1,20 +1,26 @@
...
-    public void createNewUser(@RequestBody UserDTO userDTO) {
-        userService.createNewUser(userDTO);
+    @GetMapping("/test")
+    public String test() {
+        return "{\"message\":\"Allow!\"";
+    }
+
+    @GetMapping
+    public UserDTO getUserData(@AuthenticationPrincipal User user) {
+        return userService.buildUserDtoFromUser(user);
```
Comment (TheBestTvarynka) [inline backend/src/main/java/com/eshka/controller/UserController.java]: I moved registration to AuthController

File: backend/src/main/java/com/eshka/entity/UserSession.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+@Data
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+@Entity
+@Table(name = "user_sessions")
+public class UserSession implements GrantedAuthority {
+    @Id
+    @Column(name = "id", updatable = false, nullable = false)
```
Comment (TheBestTvarynka) [inline backend/src/main/java/com/eshka/entity/UserSession.java:19]: An entity of the user session.

**Review by KySo4oK — DISMISSED** [review]

**Review by KySo4oK — APPROVED** [review]

**Pull request merged**